agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,656 @@
+	}
+
+	if a.intermediatePEM == "" {
+		return "", fmt.Errorf("secondary AWS CA provider not fully Initialized")
+	}
+
+	return a.intermediatePEM, nil
+}
+
+// GenerateIntermediate implements Provider
+func (a *AWSProvider) GenerateIntermediate() (string, error) {
+	// Like the consul provider, for now the Primary DC just gets a root and no
+	// intermediate to sign with. so just return this. Secondaries use
+	// intermediates but this method is only called during primary DC (root)
+	// initialization in case a provider generates separate root and
+	// intermediates.
+	//
+	// TODO(banks) support user-supplied CA being a Subordinate even in the
+	// primary DC.
+	return a.ActiveIntermediate()
+}
+
+// Sign implements Provider
+func (a *AWSProvider) Sign(csr *x509.CertificateRequest) (string, error) {
+	if a.rootPEM == "" {
+		return "", fmt.Errorf("AWS CA provider not fully Initialized")
+	}
+
+	a.logger.Printf("[DEBUG] connect.ca.aws: signing csr for %s",
+		csr.Subject.CommonName)
+
+	return a.signCSRRaw(csr, LeafTemplateARN, a.config.LeafCertTTL)
+}
+
+// SignIntermediate implements Provider
+func (a *AWSProvider) SignIntermediate(csr *x509.CertificateRequest) (string, error) {
+	err := validateSignIntermediate(csr, &connect.SpiffeIDSigning{ClusterID: a.clusterID, Domain: "consul"})
+	if err != nil {
+		return "", err
+	}
+
+	// Sign it!
+	return a.signCSRRaw(csr, IntermediateTemplateARN, IntermediateTTL)
+}
+
+// CrossSignCA implements Provider
+func (a *AWSProvider) CrossSignCA(newCA *x509.Certificate) (string, error) {
+	return "", fmt.Errorf("not implemented in AWS PCA provider")
+}
+
+func (a *AWSProvider) disablePCA() error {
+	if a.arn == "" {
+		return nil
+	}
+	input := acmpca.UpdateCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		Status:                  aws.String(acmpca.CertificateAuthorityStatusDisabled),
+	}
+	a.logger.Printf("[INFO] connect.ca.aws: disabling PCA %s", a.arn)
+	_, err := a.client.UpdateCertificateAuthority(&input)
+	return err
+}
+
+func (a *AWSProvider) deletePCA() error {
+	if a.arn == "" {
+		return nil
+	}
+	input := acmpca.DeleteCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		// We only ever use this to clean up after tests so delete as quickly as
+		// possible (7 days).
+		PermanentDeletionTimeInDays: aws.Int64(7),
+	}
+	a.logger.Printf("[INFO] connect.ca.aws: deleting PCA %s", a.arn)
+	_, err := a.client.DeleteCertificateAuthority(&input)
+	return err
+}
+
+// Cleanup implements Provider
+func (a *AWSProvider) Cleanup() error {
+	old := atomic.SwapUint32(&a.stopped, 1)
+	if old == 0 {
+		close(a.stopCh)
+	}
+
+	if a.config.DeleteOnExit {
+		if err := a.disablePCA(); err != nil {
+			// Log the error but continue trying to delete as some errors may still
+			// allow that and this is best-effort delete anyway.
+			a.logger.Printf("[ERR] connect.ca.aws: failed to disable PCA %s: %s",
+				a.arn, err)
+		}
+		if err := a.deletePCA(); err != nil {
```

Comment:
This couldn't be called on the new active PCA after a root rotation, right? I *think* the new root being different and triggering a rotation should mean it can't be the same PCA, but just wanted to ask.

Reply:
It should never be called outside of tests - we won't document the `DeleteOnExit` config as that is only for tests really. In prod you'd never want CA resources deleted like this right after rotation like you said.